### PR TITLE
add DDC + video datasets

### PR DESCRIPTION
**Pull request by collinleiber**
- Add DDC clustering algorithm
- Add Weizmann video dataset
- Add Keck gesture video dataset
- Change Readme layout to match #84

### CODE REVIEW

File: clustpy/data/real_video_data.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+            # Downsample frames
+            data_local, labels_local = _downsample_frames(data_local, labels_local, frame_sampling_ratio)
+            # Update data and labels
+            all_data = np.append(all_data, data_local, axis=0)
+            labels = np.append(labels, labels_local, axis=0)
+    # If desired, normalize channels
+    data_torch = torch.Tensor(all_data)
+    is_color_channel_last = True
+    data_torch = _torch_normalize_and_flatten(data_torch, flatten, normalize_channels, is_color_channel_last)
+    # Move data to CPU
```
Comment (Lumik7) [inline clustpy/data/real_video_data.py:320]: where and why did you move the data to the gpu?

File: clustpy/deep/ddc.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+         loss function for the reconstruction
+    autoencoder : torch.nn.Module
+        the input autoencoder. If None a new FeedforwardAutoencoder will be created
+    embedding_size : int
+        size of the embedding within the autoencoder
+    custom_dataloaders : tuple
+        tuple consisting of a trainloader (random order) at the first and a test loader (non-random order) at the second position.
+        If None, the default dataloaders will be used
+    tsne_params : dict
+        Parameters for the t-SNE execution. Check out sklearn.manifold.TSNE for more information
```
Comment (Lumik7) [inline clustpy/deep/ddc.py:48]: maybe add example of how perplexity value can be passed as a dictionary?

File: clustpy/deep/ddc.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    # Get initial AE
+    autoencoder = get_trained_autoencoder(trainloader, pretrain_optimizer_params, pretrain_epochs, device,
+                                          optimizer_class, loss_fn, embedding_size, autoencoder)
+    # Encode data
+    X_embed = encode_batchwise(testloader, autoencoder, device)
+    # Execute T-SNE
+    if "random_state" not in tsne_params.keys():
+        tsne_params["random_state"] = random_state
+    X_tsne = TSNE(**tsne_params).fit_transform(X_embed)
+    # Execute Density Peark Algorithm
```
Comment (Lumik7) [inline clustpy/deep/ddc.py:76]: Density Pea**r**k Algorithm --> Density Peak Algorithm

File: clustpy/deep/dec.py [hunk 4]
```diff
@@ -16,6 +17,83 @@
 from sklearn.utils import check_random_state
 
 
+def get_standard_initial_deep_clustering_setting(X: np.ndarray, n_clusters: int, batch_size: int,
```
Comment (Lumik7) [inline clustpy/deep/dec.py:20]: thats nice. Could be put in a separate utility file?

File: clustpy/deep/tests/test_ddc.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    assert np.array_equal(ddc.labels_, ddc2.labels_)
+
+
+def test_ddc_augmentation():
+    torch.use_deterministic_algorithms(True)
+    data, labels = load_optdigits(flatten=False)
+    data = data[:1000]
+    labels = labels[:1000]
+    aug_dl, orig_dl = _get_test_augmentation_dataloaders(data)
+    clusterer = DDC(pretrain_epochs=3, random_state=1,
```
Comment (Lumik7) [inline clustpy/deep/tests/test_ddc.py:28]: how does augmentation work with DDC? because I did not see any `augmentation_invariance` parameter in ddc.py

**Review by Lumik7 — APPROVED** [review]
Looks good to me. I added some minor comments. otherwise it can be merged from my side :)

File: clustpy/deep/tests/test_ddc.py (hunk 5 above)
Comment (collinleiber) [reply clustpy/deep/tests/test_ddc.py:28]: it is only used for pretraining the AE (see the original paper)

Comment (Lumik7) [reply clustpy/deep/tests/test_ddc.py:28]: ok
